File: CodeSprint-2012-03-03-Quora/FeedOptimizer/fo.cxx

#include <iostream>
#include <istream>
#include <ostream>

#include <map>
#include <set>
#include <vector>

using namespace std;

typedef signed int mint;

class Evt
{
	public:
	mint timestamp_, id_, score_, height_;

	Evt()
	{
	}

	Evt(mint timestamp, mint id, mint score, mint height):
			timestamp_(timestamp), id_(id), score_(score), height_(height)
	{
	}
};

class W
{
	public:
	mint score_, evtNum_;
	set<mint> evts_;

	W():
			score_(0), evtNum_(0)
	{
	}
};

bool operator<(const W &a, const W &b)
{
	if (a.score_ < b.score_)
	{
		return true;
	}
	else if (a.score_ > b.score_)
	{
		return false;
	}
	else
	{
		if (a.evtNum_ > b.evtNum_)
		{
			return true;
		}
		else if (a.evtNum_ < b.evtNum_)
		{
			return false;
		}
		else
		{
			set<mint>::iterator aiter = a.evts_.begin();
			set<mint>::iterator biter = b.evts_.begin();

			while (aiter != a.evts_.end())
			{
				if ((*aiter) > (*biter))
				{
					return true;
				}
				else if ((*aiter) < (*biter))
				{
					return false;
				}

				++aiter;
				++biter;
			}

			return true;
		}
	}
}

class EvtMgr
{
	public:
	mint curId_, timeWindow_, maxHeight_, stId_;
	map<mint, map<mint, W> > *m_;
	vector<Evt> evts_;
	map<mint, mint> ts_;

	EvtMgr(mint timeWindow, mint maxHeight):
			curId_(0), timeWindow_(timeWindow), maxHeight_(maxHeight), stId_(0)
	{
		m_ = new map<mint, map<mint, W> >();
	}

	void addEvt(mint timestamp, mint score, mint height)
	{
		ts_[timestamp] = curId_;
		evts_.push_back(Evt(timestamp, ++curId_, score, height));
	}

	void reload(mint timestamp)
	{
		mint curStId = stId_;

		map<mint, mint>::iterator iter = ts_.lower_bound(timestamp - timeWindow_);

		stId_ = iter->second;

		if (curStId != stId_)
		{
			delete m_;

			m_ = new map<mint, map<mint, W> >();
		}

		W w = m(curId_ - 1, maxHeight_);

		cout << w.score_ << ' ' << w.evtNum_;

		for (set<mint>::iterator witer = w.evts_.begin(); witer != w.evts_.end(); ++witer)
		{
			cout << ' ' << (*witer);
		}

		cout << endl;
	}

	W m(mint id, mint maxHeight)
	{
		if ((id < stId_) || (maxHeight <= 0))
		{
			return W();
		}

		map<mint, map<mint, W> >::iterator fiter = m_->find(id);
		
		if (fiter != m_->end())
		{
			map<mint, W>::iterator siter = fiter->second.find(maxHeight);

			if (siter != fiter->second.end())
			{
				return siter->second;
			}
		}

		if (evts_[id].height_ > maxHeight)
		{
			(*m_)[id][maxHeight] = m(id - 1, maxHeight);
		}
		else
		{
			W w = m(id - 1, maxHeight - evts_[id].height_);

			w.score_ += evts_[id].score_;
			++w.evtNum_;
			w.evts_.insert(evts_[id].id_);

			(*m_)[id][maxHeight] = max(m(id - 1, maxHeight), w);
		}

		return (*m_)[id][maxHeight];
	}
};

int main()
{
	mint numEvents, timeWindow, maxHeight;

	cin >> numEvents >> timeWindow >> maxHeight;

	EvtMgr e(timeWindow, maxHeight);

	char evtCode;

	mint timestamp, score, height;

	for (mint i = 0; i != numEvents; ++i)
	{
		cin >> evtCode;

		if (evtCode == 'S')
		{
			cin >> timestamp >> score >> height;

			e.addEvt(timestamp, score, height);
		}
		else
		{
			cin >> timestamp;

			e.reload(timestamp);
		}
	}

	return 0;
}

File: CodeSprint-2012-03-03-Quora/FeedOptimizer/fo.php
<?php

	error_reporting(E_ALL | E_STRICT);

	final class EvtMgr
	{
		private $curId = 0;

		private $timeWindow;
		private $maxHeight;

		private $m = array();
		private $stId = 1;

		private $evts = array();

		final static public function mk($timeWindow, $maxHeight)
		{
			return new self($timeWindow, $maxHeight);
		}

		final public function addEvt($timestamp, $score, $height)
		{
			++$this->curId;

			$this->evts[$this->curId] = array(
					'id' => $this->curId,
					'timestamp' => intval($timestamp),
					'score' => intval($score),
					'height' => intval($height),
					);
		}

		final public function reload($timestamp)
		{
			$ts = intval($timestamp);

			$curStId = $this->stId;

			for (; $this->evts[$this->stId]['timestamp'] < ($timestamp - $this->timeWindow); ++$this->stId)
			{
			}

			if ($curStId !== $this->stId)
			{
				$this->m = array();
			}

			return $this->m($this->curId, $this->maxHeight);
		}

		final private function __construct($timeWindow, $maxHeight)
		{
			$this->timeWindow = intval($timeWindow);
			$this->maxHeight = intval($maxHeight);
		}

		final private function m($id, $maxHeight)
		{
			if (($id < $this->stId) || ($maxHeight <= 0))
			{
				return array(
						'score' => 0,
						'numEvts' => 0,
						'evts' => array(),
						);
			}

			if ((! array_key_exists($id, $this->m)) || (! array_key_exists($maxHeight, $this->m[$id])))
			{
				if ($this->evts[$id]['height'] > $maxHeight)
				{
					$this->m[$id][$maxHeight] = $this->m($id - 1, $maxHeight);
				}
				else
				{
					$this->m[$id][$maxHeight] = $this->maxV(
							$this->m($id - 1, $maxHeight),
							$this->m($id - 1, $maxHeight - $this->evts[$id]['height']),
							$id);
				}
			}

			return $this->m[$id][$maxHeight];
		}

		final private function maxV($lst1, $lst2, $evtId)
		{
			$lst2['score'] += $this->evts[$evtId]['score'];
			++$lst2['numEvts'];
			$lst2['evts'][] = $evtId;

			sort($lst2['evts']);

			if ($lst1['score'] < $lst2['score'])
			{
				return $lst2;
			}
			elseif ($lst1['score'] > $lst2['score'])
			{
				return $lst1;
			}
			else
			{
				if ($lst1['numEvts'] < $lst2['numEvts'])
				{
					return $lst1;
				}
				elseif ($lst1['numEvts'] > $lst2['numEvts'])
				{
					return $lst2;
				}
				else
				{
					for ($i = 0; $i != count($lst1); ++$i)
					{
						if ($lst1['evts'][$i] < $lst2['evts'][$i])
						{
							return $lst1;
						}
						elseif ($lst1['evts'][$i] > $lst2['evts'][$i])
						{
							return $lst2;
						}
					}

					return $lst1;
				}
			}
		}
	}

	list($numEvents, $timeWindow, $maxHeight) = explode(' ', readline(''), 3);

	$e = EvtMgr::mk($timeWindow, $maxHeight);

	for ($i = 0; $i != $numEvents; ++$i)
	{
		$line = explode(' ', readline(''));

		$evtCode = array_shift($line);

		if ($evtCode === 'S')
		{
			list($timestamp, $score, $height) = $line;

			$e->addEvt($timestamp, $score, $height);
		}
		else
		{
			list($timestamp) = $line;

			$res = $e->reload($timestamp);

			print($res['score'].' '.$res['numEvts'].' '.implode(' ', $res['evts'])."\n");
		}
	}

?>
